File: flutter_app/lib/core/models/word.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:json_annotation/json_annotation.dart';
import 'category.dart';

part 'word.freezed.dart';
part 'word.g.dart';

@freezed
class WordListItem with _$WordListItem {
  const factory WordListItem({
    required int id,
    required String word,
    String? pronunciation,
    @JsonKey(name: 'category_id') int? categoryId,
    @JsonKey(name: 'difficulty_level') required double difficultyLevel,
    @JsonKey(name: 'importance_score') required int importanceScore,
  }) = _WordListItem;

  factory WordListItem.fromJson(Map<String, dynamic> json) =>
      _$WordListItemFromJson(json);
}

@freezed
class WordDetail with _$WordDetail {
  const factory WordDetail({
    required int id,
    required String word,
    String? pronunciation,
    @Default([]) List<String> partsOfSpeech,
    @Default([]) List<WordDefinition> definitions,
    Etymology? etymology,
    @Default([]) List<MediaItem> media,
    Category? category,
    @Default(5.0) double difficultyLevel,
    @Default(50) int importanceScore,
    required String source,
    String? tone,
    @JsonKey(name: 'cefr_level') String? cefrLevel,
  }) = _WordDetail;

  factory WordDetail.fromJson(Map<String, dynamic> json) {
    return WordDetail(
      id: json['id'] as int,
      word: json['word'] as String,
      pronunciation: json['pronunciation'] as String?,
      partsOfSpeech: (json['parts_of_speech'] as List<dynamic>?)?.cast<String>() ?? [],
      definitions: (json['definitions'] as List<dynamic>?)
          ?.map((e) => WordDefinition.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      etymology: json['etymology'] != null
          ? Etymology.fromJson(json['etymology'] as Map<String, dynamic>)
          : null,
      media: (json['media'] as List<dynamic>?)
          ?.map((e) => MediaItem.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      category: json['category'] != null
          ? Category.fromJson(json['category'] as Map<String, dynamic>)
          : null,
      difficultyLevel: (json['difficulty_level'] as num?)?.toDouble() ?? 5.0,
      importanceScore: (json['importance_score'] as num?)?.toInt() ?? 50,
      source: json['source'] as String,
      tone: json['tone'] as String?,
      cefrLevel: json['cefr_level'] as String?,
    );
  }
}

@freezed
class WordDefinition with _$WordDefinition {
  const factory WordDefinition({
    required String text,
    @Default(false) bool isPrimary,
    @Default([]) List<String> examples,
  }) = _WordDefinition;

  factory WordDefinition.fromJson(Map<String, dynamic> json) =>
      _$WordDefinitionFromJson(json);
}

@freezed
class Etymology with _$Etymology {
  const factory Etymology({
    String? originLanguage,
    String? rootWord,
    String? evolution,
  }) = _Etymology;

  factory Etymology.fromJson(Map<String, dynamic> json) =>
      _$EtymologyFromJson(json);
}

@freezed
class MediaItem with _$MediaItem {
  const factory MediaItem({
    required String type,
    required String url,
    String? source,
    String? caption,
    @Default(false) bool isAiGenerated,
  }) = _MediaItem;

  factory MediaItem.fromJson(Map<String, dynamic> json) =>
      _$MediaItemFromJson(json);
}

@freezed
class ReviewPageData with _$ReviewPageData {
  const factory ReviewPageData({
    required String word,
    String? pronunciation,
    @JsonKey(name: 'parts_of_speech') @Default([]) List<String> partsOfSpeech,
    @Default([]) List<WordDefinition> definitions,
    Etymology? etymology,
    @Default([]) List<MediaItem> media,
    Category? category,
    String? tone,
    @JsonKey(name: 'cefr_level') String? cefrLevel,
  }) = _ReviewPageData;

  factory ReviewPageData.fromJson(Map<String, dynamic> json) {
    return ReviewPageData(
      word: json['word'] as String,
      pronunciation: json['pronunciation'] as String?,
      partsOfSpeech: (json['parts_of_speech'] as List<dynamic>?)?.cast<String>() ?? [],
      definitions: (json['definitions'] as List<dynamic>?)
          ?.map((e) => WordDefinition.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      etymology: json['etymology'] != null
          ? Etymology.fromJson(json['etymology'] as Map<String, dynamic>)
          : null,
      media: (json['media'] as List<dynamic>?)
          ?.map((e) => MediaItem.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      category: json['category'] != null
          ? Category.fromJson(json['category'] as Map<String, dynamic>)
          : null,
      tone: json['tone'] as String?,
      cefrLevel: json['cefr_level'] as String?,
    );
  }
}

@freezed
class WordCreate with _$WordCreate {
  const factory WordCreate({
    required String word,
    required int categoryId,
    @Default(5.0) double difficultyLevel,
    @Default(50) int importanceScore,
    @Default([]) List<String> partOfSpeech,
    String? pronunciation,
    @Default('User') String source,
  }) = _WordCreate;

  factory WordCreate.fromJson(Map<String, dynamic> json) =>
      _$WordCreateFromJson(json);
}

File: backend/app/schemas/word.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from app.models.word_media import MediaType
from app.models.word import Tone


class WordBase(BaseModel):
    word: str
    category_id: int
    difficulty_level: float = 5.0
    importance_score: int = 50
    part_of_speech: List[str] = []
    pronunciation: Optional[str] = None
    tone: Tone = Tone.NEUTRAL
    cefr_level: Optional[str] = None  # A1, A2, B1, B2, C1, C2


class WordCreate(WordBase):
    source: str = "User"


class WordCreateSimple(BaseModel):
    """Simple word creation - only requires the word, AI generates the rest."""
    word: str


class WordDefinitionSchema(BaseModel):
    text: str
    is_primary: bool
    examples: List[str] = []


class EtymologySchema(BaseModel):
    origin_language: Optional[str] = None
    root_word: Optional[str] = None
    evolution: Optional[str] = None


class MediaSchema(BaseModel):
    type: str
    url: str
    source: Optional[str] = None
    caption: Optional[str] = None
    is_ai_generated: bool = False


class MediaCreate(BaseModel):
    """Request schema for creating media."""
    word_id: int
    media_type: MediaType
    url: str
    source: Optional[str] = None
    caption: Optional[str] = None
    is_ai_generated: bool = False


class CategorySchema(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    parent_category_id: Optional[int] = None
    importance_weight: float = 5.0

    class Config:
        from_attributes = True


class WordDetail(BaseModel):
    id: int
    word: str
    pronunciation: Optional[str] = None
    parts_of_speech: List[str] = []
    definitions: List[WordDefinitionSchema] = []
    etymology: Optional[EtymologySchema] = None
    media: List[MediaSchema] = []
    category: Optional[CategorySchema] = None
    difficulty_level: float
    importance_score: int
    source: str
    tone: str
    cefr_level: Optional[str] = None

    class Config:
        from_attributes = True


class WordListItem(BaseModel):
    id: int
    word: str
    pronunciation: Optional[str] = None
    category_id: int
    difficulty_level: float
    importance_score: int
    tone: str
    cefr_level: Optional[str] = None

    class Config:
        from_attributes = True


class ReviewPageData(BaseModel):
    word: str
    pronunciation: Optional[str] = None
    parts_of_speech: List[str] = []
    definitions: List[WordDefinitionSchema] = []
    etymology: Optional[EtymologySchema] = None
    media: List[MediaSchema] = []
    category: Optional[CategorySchema] = None
    tone: str
    cefr_level: Optional[str] = None